src/modules/rlm_ldap/user.c

```diff
@@ -35,6 +35,136 @@ USES_APPLE_DEPRECATED_API
 
 #include "rlm_ldap.h"
 
+/** Holds state of user searches in progress
+ *
+ */
+typedef struct {
+	rlm_ldap_t const	*inst;
+	fr_ldap_thread_trunk_t	*ttrunk;
+	char const		*base_dn;
+	char const		*filter;
+	char const * const	*attrs;
+	fr_ldap_query_t		*query;
+	fr_ldap_query_t		**out;
+} ldap_user_find_ctx;
+
+static unlang_action_t ldap_find_user_async_result(rlm_rcode_t *p_result, UNUSED int *priority, request_t *request, void *uctx)
+{
+	ldap_user_find_ctx	*user_ctx = talloc_get_type_abort(uctx, ldap_user_find_ctx);
+	fr_ldap_query_t		*query = user_ctx->query;
+	LDAPMessage		*entry;
+	int			cnt, ldap_errno;
+	char			*dn;
+	fr_pair_t		*vp;
+
+	if (!user_ctx->inst->userobj_sort_ctrl) {
+		cnt = ldap_count_entries(query->ldap_conn->handle, query->result);
+		if (cnt > 1) {
+			REDEBUG("Ambiguous search result, returned %i unsorted entries (should return 1 or 0).  "
+				"Enable sorting, or specify a more restrictive base_dn, filter or scope", cnt);
+			REDEBUG("The following entries were returned:");
+			RINDENT();
+			for (entry = ldap_first_entry(query->ldap_conn->handle, query->result);
+			     entry;
+			     entry = ldap_next_entry(query->ldap_conn->handle, entry)) {
+				dn = ldap_get_dn(query->ldap_conn->handle, entry);
+				REDEBUG("%s", dn);
+				ldap_memfree(dn);
+			}
+			REXDENT();
+			RETURN_MODULE_INVALID;
+		}
+	}
+
+	entry = ldap_first_entry(query->ldap_conn->handle, query->result);
+	if (!entry) {
```

Comment:
Just to make sure... There's a separate code path which deals with users that weren't found, right?  Else this should be `RETURN_MODULE_NOTFOUND`